File: cfms_bd/app/Http/Controllers/Admin/SessionController.php
<?php

namespace App\Http\Controllers\Admin;

use App\Http\Controllers\Controller;
use App\Models\Session;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Validator;

class SessionController extends Controller
{
    /**
     * 1. Display a listing of all sessions.
     */
    public function index()
    {
        try {
            $sessions = Session::orderBy('id', 'desc')->get();
            
            return response()->json([
                'success' => true,
                'data' => $sessions
            ], 200);
        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'Database error occurred while fetching sessions: ' . $e->getMessage()
            ], 500);
        }
    }

    /**
     * 2. Store a newly created session in the database.
     */
    public function store(Request $request)
    {
        $validator = Validator::make($request->all(), [
            's_name'     => 'required|string|max:255',
            'start_date' => 'required|date_format:d/m/Y',
            'end_date'   => 'required|date_format:d/m/Y|after_or_equal:start_date',
        ]);

        if ($validator->fails()) {
            return response()->json([
                'success' => false,
                'message' => 'Validation errors occurred.',
                'errors'  => $validator->errors()
            ], 422);
        }

        try {
            $session = Session::create([
                's_name'     => $request->s_name,
                'start_date' => $request->start_date,
                'end_date'   => $request->end_date,
            ]);

            return response()->json([
                'success' => true,
                'message' => 'Session created successfully.',
                'data'    => $session
            ], 201);
        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'Failed to create session: ' . $e->getMessage()
            ], 500);
        }
    }

    /**
     * 3. Display the specified session resource.
     */
    public function show($id)
    {
        $session = Session::find($id);

        if (!$session) {
            return response()->json([
                'success' => false,
                'message' => 'Requested session record not found.'
            ], 404);
        }

        return response()->json([
            'success' => true,
            'data'    => $session
        ], 200);
    }

    /**
     * 4. Update the specified session inside the database.
     */
    public function update(Request $request, $id)
    {
        $session = Session::find($id);

        if (!$session) {
            return response()->json([
                'success' => false,
                'message' => 'Session not found to perform modification.'
            ], 404);
        }

        $validator = Validator::make($request->all(), [
            's_name'     => 'required|string|max:255',
            'start_date' => 'required|date_format:d/m/Y',
            'end_date'   => 'required|date_format:d/m/Y|after_or_equal:start_date',
        ]);

        if ($validator->fails()) {
            return response()->json([
                'success' => false,
                'message' => 'Validation errors occurred.',
                'errors'  => $validator->errors()
            ], 422);
        }

        try {
            $session->update([
                's_name'     => $request->s_name,
                'start_date' => $request->start_date,
                'end_date'   => $request->end_date,
            ]);

            return response()->json([
                'success' => true,
                'message' => 'Session updated successfully.',
                'data'    => $session
            ], 200);
        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'Failed to update session data: ' . $e->getMessage()
            ], 500);
        }
    }

    /**
     * 5. Remove the specified session from storage.
     */
    public function destroy($id)
    {
        $session = Session::find($id);

        if (!$session) {
            return response()->json([
                'success' => false,
                'message' => 'Session record already missing or deleted.'
            ], 404);
        }

        try {
            $session->delete();

            return response()->json([
                'success' => true,
                'message' => 'Session record removed successfully from database.'
            ], 200);
        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'System error occurred during record deletion: ' . $e->getMessage()
            ], 500);
        }
    }
}

File: cfms_fd/lib/ViewModel/Controller/AdminController/SessionController.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class SessionController {
  final String baseUrl = 'http://127.0.0.1:8000/api';

  // 1. Fetch All Sessions
  Future<Map<String, dynamic>> fetchSessions(String token) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/admin/sessions'),
        headers: {
          'Authorization': 'Bearer $token',
          'Accept': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final decodedData = json.decode(response.body);
        if (decodedData is List) {
          return {'success': true, 'data': decodedData};
        } else if (decodedData is Map<String, dynamic>) {
          return {
            'success': decodedData['success'] ?? true,
            'data': decodedData['data'] ?? [],
          };
        }
        return {'success': true, 'data': decodedData};
      } else {
        return {'success': false, 'message': 'Failed to load sessions from server.'};
      }
    } catch (e) {
      return {'success': false, 'message': 'API Database Error: $e'};
    }
  }

  // 2. Create New Session (Sends direct DD/MM/YYYY)
  Future<Map<String, dynamic>> createSession(String token, String sName, String startDate, String endDate) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/admin/sessions'),
        headers: {
          'Authorization': 'Bearer $token',
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: json.encode({
          's_name': sName,
          'start_date': startDate, // Already in dd/mm/yyyy
          'end_date': endDate,     // Already in dd/mm/yyyy
        }),
      );

      final data = json.decode(response.body);

      if (response.statusCode == 201 || response.statusCode == 200) {
        return {'success': true, 'message': 'Session created successfully.', 'data': data['data']};
      } else if (response.statusCode == 422) {
        return {'success': false, 'message': data['message'] ?? 'Validation errors occurred.'};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Failed to create session.'};
      }
    } catch (e) {
      return {'success': false, 'message': 'API Error: $e'};
    }
  }

  // 3. Update Existing Session
  Future<Map<String, dynamic>> updateSession(String token, int id, String sName, String startDate, String endDate) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl/admin/sessions/$id'),
        headers: {
          'Authorization': 'Bearer $token',
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: json.encode({
          's_name': sName,
          'start_date': startDate,
          'end_date': endDate,
        }),
      );

      final data = json.decode(response.body);

      if (response.statusCode == 200) {
        return {'success': true, 'message': 'Session updated successfully.'};
      } else if (response.statusCode == 422) {
        return {'success': false, 'message': data['message'] ?? 'Validation errors occurred.'};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Failed to update session.'};
      }
    } catch (e) {
      return {'success': false, 'message': 'API Error: $e'};
    }
  }

  // 4. Delete Session
  Future<Map<String, dynamic>> deleteSession(String token, int id) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/admin/sessions/$id'),
        headers: {
          'Authorization': 'Bearer $token',
          'Accept': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        return {'success': true, 'message': 'Session deleted successfully.'};
      } else {
        return {'success': false, 'message': 'Failed to delete session.'};
      }
    } catch (e) {
      return {'success': false, 'message': 'API Error: $e'};
    }
  }
}